bridges/snowbridge/pallets/outbound-queue-v2/src/lib.rs

```diff
@@ -0,0 +1,402 @@
+		/// Invalid Gateway
+		InvalidGateway,
+		/// Pending nonce does not exist
+		InvalidPendingNonce,
+		/// Reward payment failed
+		RewardPaymentFailed,
+	}
+
+	/// Messages to be committed in the current block. This storage value is killed in
+	/// `on_initialize`, so will not end up bloating state.
+	///
+	/// Is never read in the runtime, only by offchain message relayers.
+	/// Because of this, it will never go into the PoV of a block.
+	///
+	/// Inspired by the `frame_system::Pallet::Events` storage value
+	#[pallet::storage]
+	#[pallet::unbounded]
+	pub(super) type Messages<T: Config> = StorageValue<_, Vec<OutboundMessage>, ValueQuery>;
+
+	/// Hashes of the ABI-encoded messages in the [`Messages`] storage value. Used to generate a
+	/// merkle root during `on_finalize`. This storage value is killed in
+	/// `on_initialize`, so should never go into block PoV.
+	#[pallet::storage]
+	#[pallet::unbounded]
+	pub(super) type MessageLeaves<T: Config> = StorageValue<_, Vec<H256>, ValueQuery>;
+
+	/// The current nonce for the messages
+	#[pallet::storage]
+	pub type Nonce<T: Config> = StorageValue<_, u64, ValueQuery>;
+
+	/// Pending orders to relay
+	#[pallet::storage]
+	pub type PendingOrders<T: Config> =
+		StorageMap<_, Identity, u64, PendingOrder<BlockNumberFor<T>>, OptionQuery>;
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T>
+	where
+		T::AccountId: AsRef<[u8]>,
+	{
+		fn on_initialize(_: BlockNumberFor<T>) -> Weight {
+			// Remove storage from previous block
+			Messages::<T>::kill();
+			MessageLeaves::<T>::kill();
+			// Reserve some weight for the `on_finalize` handler
+			T::WeightInfo::commit()
+		}
+
+		fn on_finalize(_: BlockNumberFor<T>) {
+			Self::commit();
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		T::AccountId: From<[u8; 32]>,
+	{
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::submit_delivery_receipt())]
+		pub fn submit_delivery_receipt(
+			origin: OriginFor<T>,
+			event: Box<EventProof>,
+		) -> DispatchResult {
+			let relayer = ensure_signed(origin)?;
+
+			// submit message to verifier for verification
+			T::Verifier::verify(&event.event_log, &event.proof)
+				.map_err(|e| Error::<T>::Verification(e))?;
+
+			let receipt = DeliveryReceiptOf::<T>::try_from(&event.event_log)
+				.map_err(|_| Error::<T>::InvalidEnvelope)?;
+
+			Self::process_delivery_receipt(relayer, receipt)
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Generate a messages commitment and insert it into the header digest
+		pub(crate) fn commit() {
+			let count = MessageLeaves::<T>::decode_len().unwrap_or_default() as u64;
+			if count == 0 {
+				return;
+			}
+
+			// Create merkle root of messages
+			let root = merkle_root::<<T as Config>::Hashing, _>(MessageLeaves::<T>::stream_iter());
+
+			let digest_item: DigestItem = CustomDigestItem::SnowbridgeV2(root).into();
+
+			// Insert merkle root into the header digest
+			<frame_system::Pallet<T>>::deposit_log(digest_item);
```

Comment:
can we kill `MessageLeaves` here so that it never makes it to storage? it can be rebuilt in the generate_proof runtime-api call from `Messages`.
```suggestion
			<frame_system::Pallet<T>>::deposit_log(digest_item);
			MessageLeaves::<T>::kill();
```

Reply:
@vgeddes @yrong wdyt about this? it's not mandatory, but a storage optimization

Reply:
opened https://github.com/paritytech/polkadot-sdk/issues/7971 for following it up